### PR TITLE
[LoRA] Fix LoRA merge and support CanonicalLoRA merge

### PR DESCRIPTION
**Pull request by HollowMan6**
# What does this PR do ?

Previous LoRA merge #1310 contains several errors, this PR aims to fix it and support CanonicalLoRA merge

# Changelog

- Handle fused QKV/gate up correctly
- Handle PP gathering correctly

# GitHub Actions CI

See the [CI section](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md#-running-github-ci)in the Contributing doc for how to trigger the CI. A Nvidia developer will need to approve and trigger the CI for external contributors.

# Before your PR is "Ready for review"

**Pre checks**:

- [X] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md)
- [X] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?

If you haven't finished some of the above items you can still open "Draft" PR.

# Additional Information

- Related to #1310

<sub>✨ Presented to you with <a href="https://macaron.im/mindlab">Mind Lab</a> - A Lab for Experiential Intelligence.</sub>

### CODE REVIEW

**Comment by HollowMan6** [comment]
cc: @yaoyu-33 @cuichenx @ananthsub

https://github.com/volcengine/verl/pull/4063#issuecomment-3616423101

File: src/megatron/bridge/models/conversion/param_mapping.py [hunk 1]
```diff
@@ -852,7 +852,7 @@ def megatron_to_hf(
             gathered = self.gather_from_tp_ranks(megatron_weights)
             full_weights = torch.cat(gathered, dim=0)
 
-        if self.is_expert:
+        if self.is_expert and not ".adapter." in self.megatron_param:
```
Comment (yaoyu-33) [inline src/megatron/bridge/models/conversion/param_mapping.py]: can you wrap it up. use self.is_adapter? so we can change the criteria later if needed. judge if adapter based on naming can be a bit rough.

Comment (yaoyu-33) [reply src/megatron/bridge/models/conversion/param_mapping.py]: if you have similar condition in other place please adjust as well.

File: src/megatron/bridge/models/conversion/model_bridge.py [hunk 2]
```diff
@@ -136,19 +144,14 @@ def _megatron_local_name_to_global(
...
         def _update_expert_number(param_name: str, param_type: str) -> str:
             """Update expert number from local to global for weight or bias parameters."""
-            suffix = param_name.split(f".{param_type}")[-1]
-            # Check if suffix contains a valid expert number
-            # (this can be missing from PEFT adapters weight)
-            if not suffix or not suffix.isdigit():
-                # No expert number suffix, return original param_name
-                return param_name
-            local_expert_number = int(suffix)
+            local_expert_number = int(param_name.split(f".{param_type}")[-1])
```
Comment (yaoyu-33) [inline src/megatron/bridge/models/conversion/model_bridge.py:196]: why need to change?

File: src/megatron/bridge/models/conversion/model_bridge.py [hunk 3]
```diff
@@ -295,18 +298,31 @@ def mapping_registry(self):
...
+        adapter = getattr(lora_module, "adapter", None)
+        if adapter is None:
+            # For CanonicalLoRA module
+            lora_module, _ = get_module_and_param_from_name(megatron_model, base_name + ".to_wrap", vp_stage)
+        return getattr(lora_module, "adapter", None), getattr(lora_module, "to_wrap", None)
+
     def _megatron_global_param_names_all_pp_ranks(
-        self, megatron_model: Union[MegatronModel, List[MegatronModel]]
-    ) -> List[str]:
+        self, megatron_model: Union[MegatronModel, List[MegatronModel]], for_adapter: bool = False
```
Comment (yaoyu-33) [inline src/megatron/bridge/models/conversion/model_bridge.py]: feels can be 2 methods. just create another method for adapter.

Comment (yaoyu-33) [reply src/megatron/bridge/models/conversion/model_bridge.py]: just split the adapter logic, if it needs a object to describe.

File: src/megatron/bridge/models/conversion/model_bridge.py (hunk 2 above)
Comment (HollowMan6) [reply src/megatron/bridge/models/conversion/model_bridge.py:196]: This is for reverting the changes in the last PR: https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/1310/files#diff-9ed67c2305fea62d01673808de1a78b6b41ba9d9f50a2810f242da25258ffd0cL144-R150 As I found that it's not necessary to have this when we already have https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/1603#discussion_r2593439494, as adapters are not EP sharded anyway.

File: src/megatron/bridge/models/conversion/param_mapping.py (hunk 1 above)
Comment (HollowMan6) [reply src/megatron/bridge/models/conversion/param_mapping.py]: Done!

File: src/megatron/bridge/models/conversion/model_bridge.py (hunk 3 above)
Comment (HollowMan6) [reply src/megatron/bridge/models/conversion/model_bridge.py]: Done

File: src/megatron/bridge/models/conversion/model_bridge.py [hunk 4]
```diff
@@ -341,6 +355,67 @@ def _megatron_global_param_names_all_pp_ranks(
 
         return self._cached_param_names
 
+    def _megatron_global_adapter_infos_all_pp_ranks(
```
Comment (yaoyu-33) [inline src/megatron/bridge/models/conversion/model_bridge.py]: info, no "s", lol

File: src/megatron/bridge/models/conversion/model_bridge.py [hunk 5]
```diff
@@ -341,6 +355,67 @@ def _megatron_global_param_names_all_pp_ranks(
...
+        self, megatron_model: Union[MegatronModel, List[MegatronModel]]
+    ) -> List[tuple[str, bool, bool, int, int]]:
+        """Get all adapter parameter names across all pipeline parallel ranks."""
```
Comment (yaoyu-33) [inline src/megatron/bridge/models/conversion/model_bridge.py]: fix docstring, not only names

File: src/megatron/bridge/models/conversion/model_bridge.py (hunk 4 above)
Comment (HollowMan6) [reply src/megatron/bridge/models/conversion/model_bridge.py]: Here I just want to emphasize this is information about a list of adapters, instead of a single adapter, XD, maybe I'll just change all `adapter_infos` to `adapters_info`. I have several `adapter_infos` for the rest of the code as well XD.

File: src/megatron/bridge/models/conversion/model_bridge.py (hunk 5 above)
Comment (HollowMan6) [reply src/megatron/bridge/models/conversion/model_bridge.py]: Ah, forgot about that, good catch!

Comment (HollowMan6) [reply src/megatron/bridge/models/conversion/model_bridge.py]: done

File: src/megatron/bridge/models/conversion/model_bridge.py (hunk 4 above)
Comment (HollowMan6) [reply src/megatron/bridge/models/conversion/model_bridge.py]: done

**Comment by HollowMan6** [comment]
Now this PR is refactored for better support of future LoRA bridge, we first construct the `AdapterWeightConversionTask` using the information from `_megatron_global_adapters_info_all_pp_ranks()` and build a dict using `global_base_prefix` as key for indexing those tasks, then we materialize (finish) those adapters' tasks with base weight conversion tasks in the streaming manner, and store them into a list `AdapterWeight`. Finally we do LoRA merge using those `AdapterWeight`.

File: src/megatron/bridge/models/conversion/model_bridge.py [hunk 6]
```diff
@@ -63,6 +73,14 @@
 MegatronModel = TypeVar("MegatronModel", bound=MegatronModule)
 _BridgeImplClass = TypeVar("_BridgeImplClass", bound="MegatronModelBridge")
 
+ADAPTER_NAME_MAP = {
```
Comment (yaoyu-33) [inline src/megatron/bridge/models/conversion/model_bridge.py:76]: need comment, what's left and right.

File: src/megatron/bridge/models/conversion/model_bridge.py [hunk 7]
```diff
@@ -104,13 +123,43 @@ class WeightConversionTask(Generic[MappingT]):
...
+    global_base_prefix: str
+    adapter_key: Optional[str]  # For canonical LoRA only
+    alpha: int
+    dim: int
+    linear_in_task: WeightConversionTask
+    linear_out_task: WeightConversionTask
+
+
+@dataclass(frozen=True)
+class AdapterWeight:
```
Comment (yaoyu-33) [inline src/megatron/bridge/models/conversion/model_bridge.py:153]: quite a few duplicated here and above. are they both needed. is it something close to MegatronWeightTuple(NamedTuple):?

File: src/megatron/bridge/models/conversion/model_bridge.py [hunk 8]
```diff
@@ -295,6 +339,17 @@ def mapping_registry(self):
         """
         raise NotImplementedError("Subclass must implement mapping_registry method")
 
+    def _get_adapter_wrap_module(
```
Comment (yaoyu-33) [inline src/megatron/bridge/models/conversion/model_bridge.py:343]: missing docstring

File: src/megatron/bridge/models/conversion/model_bridge.py [hunk 9]
```diff
@@ -341,6 +396,207 @@ def _megatron_global_param_names_all_pp_ranks(
...
+
+        # the order cannot be changed, this sync for all ranks for conversion
+        # change this might cause a hang
+        gathered_global_param_objects = sorted(flattened_names, key=lambda x: extract_sort_key(x[0]))
+
+        self._cached_param_objects_adapter = gathered_global_param_objects
+
+        return gathered_global_param_objects
+
+    def _construct_adapters_names(self, prefix: str, adapter_key: Optional[str]) -> tuple[str, str]:
```
Comment (yaoyu-33) [inline src/megatron/bridge/models/conversion/model_bridge.py:477]: missing docstring

File: src/megatron/bridge/models/conversion/model_bridge.py (hunk 7 above)
Comment (yaoyu-33) [reply src/megatron/bridge/models/conversion/model_bridge.py:153]: it's okay to duplicated, just confirm, i see you passed in only weight in merge.

Comment (HollowMan6) [reply src/megatron/bridge/models/conversion/model_bridge.py:153]: Yes, both of them are needed for abstraction. The difference between `AdapterWeightConversionTask` and `AdapterWeight` is that the former one is "unmaterialized" tasks and the latter is the matterialized weight. During merge we do need alpha and dim (as args to pass for `_merge_single_adapter_weight`), and `adapter_key` is needed to map the hf params with the adapters. `global_base_prefix` is for later use if we want to have some separate logic for LoRA bridge. Those informations are originally from `_megatron_global_adapters_info_all_pp_ranks` and we only pass those to `AdapterWeightConversionTask` in `build_adapter_conversion_tasks`, and in `materialize_adapter_weights` we pass those information from `AdapterWeightConversionTask` to `AdapterWeight`.

Comment (HollowMan6) [reply src/megatron/bridge/models/conversion/model_bridge.py:153]: > is it something close to MegatronWeightTuple(NamedTuple)

Yeah, I guess I can turn the `linear_in_weight` and `linear_out_weight` properties of `AdapterWeight` into `MegatronWeightTuple` instead of just `torch.Tensor` for better abstraction.

Comment (HollowMan6) [reply src/megatron/bridge/models/conversion/model_bridge.py:153]: > Yeah, I guess I can turn the `linear_in_weight` and `linear_out_weight` properties of `AdapterWeight` into `MegatronWeightTuple` instead of just `torch.Tensor` for better abstraction.

Done

File: src/megatron/bridge/models/conversion/model_bridge.py (hunk 9 above)
Comment (HollowMan6) [reply src/megatron/bridge/models/conversion/model_bridge.py:477]: added

File: src/megatron/bridge/models/conversion/model_bridge.py (hunk 8 above)
Comment (HollowMan6) [reply src/megatron/bridge/models/conversion/model_bridge.py:343]: added

File: src/megatron/bridge/models/conversion/model_bridge.py (hunk 6 above)
Comment (HollowMan6) [reply src/megatron/bridge/models/conversion/model_bridge.py:76]: added

**Comment by HollowMan6** [comment]
Test results on MoE (Qwen3-30B-A3B) LoRA also looks good:

image

Canonical LoRA will work on MoE with the following PRs merged:
- #1627 
- #1628

**Review by yaoyu-33 — APPROVED** [review]

**Comment by HollowMan6** [comment]
Looks like the CI error is triggered by some timeout issue: https://github.com/NVIDIA-NeMo/Megatron-Bridge/actions/runs/20009063590/job/57406565605?pr=1603, would you mind triggering it again?
